File: tzio-core/scala/src/main/scala/com/kineolyan/tzio/v1/scala/operations/OperationAdapter.scala
package com.kineolyan.tzio.v1.scala.operations

import com.kineolyan.tzio.v1.api._
import com.kineolyan.tzio.v1.scala.refs.{InputAdapter, OutputAdapter}

/**
  * Object adapting the definition of an operation to their actual types in the Scala core
  */
object OperationAdapter extends ops.OperationVisitor[Operation] {

  def convert(definition: ops.OperationType): Operation =
    definition.accept(this)

  override def visit(movOperation: ops.MovOperation): Operation =
    MovOperation(
      InputAdapter.convert(movOperation.input),
      OutputAdapter.convert(movOperation.output))

  override def visit(savOperation: ops.SavOperation): Operation =
    SavOperation(savOperation.slot)

  override def visit(swpOperation: ops.SwpOperation): Operation =
    SwpOperation(swpOperation.slot)

  override def visit(addOperation: ops.AddOperation): Operation =
    AddOperation(
      InputAdapter.convert(addOperation.input))

  override def visit(subOperation: ops.SubOperation): Operation =
    SubOperation(
      InputAdapter.convert(subOperation.input))

  override def visit(negOperation: ops.NegOperation): Operation = NegOperation()

  override def visit(labelOperation: ops.LabelOperation): Operation =
    LblOperation(labelOperation.label)

  override def visit(jmpOperation: ops.JmpOperation): Operation =
    JmpOperation(jmpOperation.label)

  override def visit(jezOperation: ops.JezOperation): Operation =
    JezOperation(jezOperation.label)

  override def visit(jnzOperation: ops.JnzOperation): Operation =
    JnzOperation(jnzOperation.label)

  override def visit(jlzOperation: ops.JlzOperation): Operation =
    JlzOperation(jlzOperation.label)

  override def visit(jgzOperation: ops.JgzOperation): Operation =
    JgzOperation(jgzOperation.label)

  override def visit(jroOperation: ops.JroOperation): Operation =
    JroOperation(
      InputAdapter.convert(jroOperation.input))
}

File: tzio-core/jdk18/src/main/java/com/kineolyan/tzio/v1/java/ops/OperationAdapter.java
package com.kineolyan.tzio.v1.java.ops;

import com.kineolyan.tzio.v1.api.ops.AddOperation;
import com.kineolyan.tzio.v1.api.ops.JezOperation;
import com.kineolyan.tzio.v1.api.ops.JgzOperation;
import com.kineolyan.tzio.v1.api.ops.JlzOperation;
import com.kineolyan.tzio.v1.api.ops.JmpOperation;
import com.kineolyan.tzio.v1.api.ops.JnzOperation;
import com.kineolyan.tzio.v1.api.ops.JroOperation;
import com.kineolyan.tzio.v1.api.ops.LabelOperation;
import com.kineolyan.tzio.v1.api.ops.MovOperation;
import com.kineolyan.tzio.v1.api.ops.NegOperation;
import com.kineolyan.tzio.v1.api.ops.OperationType;
import com.kineolyan.tzio.v1.api.ops.OperationVisitor;
import com.kineolyan.tzio.v1.api.ops.SavOperation;
import com.kineolyan.tzio.v1.api.ops.SubOperation;
import com.kineolyan.tzio.v1.api.ops.SwpOperation;
import com.kineolyan.tzio.v1.java.ref.InputAdapter;
import com.kineolyan.tzio.v1.java.ref.OutputAdapter;
import lombok.RequiredArgsConstructor;

/**
 * Adapter converting operation description into their implementation for this core.
 */
@RequiredArgsConstructor
public class OperationAdapter implements OperationVisitor<Operation> {

	/** Adapter to use for inputs */
	private final InputAdapter inputAdapter;
	/** Adapter to use for outputs */
	private final OutputAdapter outputAdapter;

	public Operation convert(OperationType type) {
		return type.accept(this);
	}

	@Override
	public Operation visit(final MovOperation movOperation) {
		return Operations.MOV(
			this.inputAdapter.convert(movOperation.input()),
			this.outputAdapter.convert(movOperation.output()));
	}

	@Override
	public Operation visit(final SavOperation savOperation) {
		return Operations.SAV(savOperation.slot());
	}

	@Override
	public Operation visit(final SwpOperation swpOperation) {
		return Operations.SWP(swpOperation.slot());
	}

	@Override
	public Operation visit(final AddOperation addOperation) {
		return Operations.ADD(
				this.inputAdapter.convert(addOperation.input()));
	}

	@Override
	public Operation visit(final SubOperation subOperation) {
		return Operations.SUB(
				this.inputAdapter.convert(subOperation.input()));
	}

	@Override
	public Operation visit(final NegOperation negOperation) {
		return Operations.NEG();
	}

	@Override
	public Operation visit(final LabelOperation labelOperation) {
		return Operations.LABEL(labelOperation.label());
	}

	@Override
	public Operation visit(final JmpOperation jmpOperation) {
		return Operations.JMP(jmpOperation.label());
	}

	@Override
	public Operation visit(final JezOperation jezOperation) {
		return Operations.JEZ(jezOperation.label());
	}

	@Override
	public Operation visit(final JnzOperation jnzOperation) {
		return Operations.JNZ(jnzOperation.label());
	}

	@Override
	public Operation visit(final JlzOperation jlzOperation) {
		return Operations.JLZ(jlzOperation.label());
	}

	@Override
	public Operation visit(final JgzOperation jgzOperation) {
		return Operations.JGZ(jgzOperation.label());
	}

	@Override
	public Operation visit(final JroOperation jroOperation) {
		return Operations.JRO(
				this.inputAdapter.convert(jroOperation.input()));
	}

}
